### ISSUE TITLE
Transition from github.com/gogo/protobuf to google.golang.org/protobuf

### ISSUE THREAD

**Issue by yoheiueda** [kata-containers/kata-containers#7420]
`github.com/gogo/protobuf`has been deprecated.

Dependency to `gogo/protobuf` has been already removed from containred and cri-o.

https://github.com/containerd/containerd/issues/6564
https://github.com/containerd/ttrpc/pull/99
https://github.com/cri-o/cri-o/issues/6350#issuecomment-1309988197
https://github.com/cri-o/cri-o/pull/6382

Kata still has dependency to `gogo/protobuf`, which causes incompatibility with external modules that uses the latest `google.golang.org/protobuf` like this one.

https://github.com/confidential-containers/cloud-api-adaptor/issues/1212

One of the reason of the incompatibility is the fact that an old version of the `genproto` package needs to be used with `gogo/protobuf` by using the `replace` directive in go.mod.

https://github.com/kata-containers/kata-containers/blob/e1a4040a6c810b84df29ee71615b35b49fe0f46c/src/runtime/go.mod#L116

**Comment by yoheiueda** [comment]
I did a quick research on how to migrate from `github.com/gogo/protobuf` to `google.golang.org/protobuf`, but it turns out to be complicated, and we need to update a lot of code and scripts.

I listed what I found so far. I don't think this list is exhaustive. It looks like Kata extensively uses protobuf features. I am not an export of this area, and I think we need some volunteers to fix this issue.

### 1. Remove the replace directive for the `genproto` package in go.mod

https://github.com/kata-containers/kata-containers/blob/e1a4040a6c810b84df29ee71615b35b49fe0f46c/src/runtime/go.mod#L116

### 2. Update the references to containred modules in go.mod
https://github.com/kata-containers/kata-containers/blob/e1a4040a6c810b84df29ee71615b35b49fe0f46c/src/runtime/go.mod#L12-L15

### 3. Update the script that generates protobuf go source files.

  [src/libs/protocols/hack/update-generated-proto.sh](https://github.com/kata-containers/kata-containers/blob/e1a4040a6c810b84df29ee71615b35b49fe0f46c/src/libs/protocols/hack/update-generated-proto.sh#L50-L62)

It may be a good time to migrate from `protoc` to [protobiuld](https://github.com/containerd/protobuild).

We can still use `protoc` by installing [protoc-gen-go-ttrpc](https://github.com/containerd/ttrpc/tree/main/cmd/protoc-gen-go-ttrpc). It looks like we need to specify both `--go_out` and `--go-ttrpc_out` options to `protoc` like this.
```
protoc -I$GOPATH/src:$GOPATH/src/github.com/kata-containers/kata-containers/src/libs/protocols/protos \
	--go_out=Mgithub.com/kata-containers/kata-containers/src/libs/protocols/protos/csi.proto=github.com/kata-containers/kata-containers/src/runtime/virtcontainers/pkg/agent/protocols/grpc,Mgithub.com/kata-containers/kata-containers/src/libs/protocols/protos/types.proto=github.com/kata-containers/kata-containers/src/runtime/virtcontainers/pkg/agent/protocols,Mgithub.com/kata-containers/kata-containers/src/libs/protocols/protos/oci.proto=github.com/kata-containers/kata-containers/src/runtime/virtcontainers/pkg/agent/protocols/grpc,Mgogoproto/gogo.proto=github.com/gogo/protobuf/gogoproto,Mgoogle/protobuf/any.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/descriptor.proto=github.com/gogo/protobuf/protoc-gen-gogo/descriptor,Mgoogle/protobuf/duration.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/empty.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/field_mask.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/timestamp.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/wrappers.proto=github.com/gogo/protobuf/types,Mgoogle/rpc/status.proto=github.com/gogo/googleapis/google/rpc:$GOPATH/src \
	--go-ttrpc_out=plugins=fieldpath,import_path=github.com/kata-containers/kata-containers/src/runtime/virtcontainers/pkg/agent/protocols/grpc,Mgithub.com/kata-containers/kata-containers/src/libs/protocols/protos/csi.proto=github.com/kata-containers/kata-containers/src/runtime/virtcontainers/pkg/agent/protocols/grpc,Mgithub.com/kata-containers/kata-containers/src/libs/protocols/protos/types.proto=github.com/kata-containers/kata-containers/src/runtime/virtcontainers/pkg/agent/protocols,Mgithub.com/kata-containers/kata-containers/src/libs/protocols/protos/oci.proto=github.com/kata-containers/kata-containers/src/runtime/virtcontainers/pkg/agent/protocols/grpc,Mgogoproto/gogo.proto=github.com/gogo/protobuf/gogoproto,Mgoogle/protobuf/any.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/descriptor.proto=github.com/gogo/protobuf/protoc-gen-gogo/descriptor,Mgoogle/protobuf/duration.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/empty.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/field_mask.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/timestamp.proto=github.com/gogo/protobuf/types,Mgoogle/protobuf/wrappers.proto=github.com/gogo/protobuf/types,Mgoogle/rpc/status.proto=github.com/gogo/googleapis/google/rpc:$GOPATH/src \
	$GOPATH/src/github.com/kata-containers/kata-containers/src/libs/protocols/protos/$1"
```


### 4. Update the reference of the task API
```
taskAPI "github.com/containerd/containerd/runtime/v2/task"
```
```
taskAPI "github.com/containerd/containerd/api/runtime/task/v2"
```

### 5 Remove uses of gogo specific types

Remove this kind of direct uses of gogo.
https://github.com/kata-containers/kata-containers/blob/e1a4040a6c810b84df29ee71615b35b49fe0f46c/src/libs/protocols/protos/oci.proto#L12

### 6. Update CCv0

Migration from gogo in CCv0 is much complicated, since it depends on [the forked version of containred](https://github.com/confidential-containers/containerd/tree/CC-main).

https://github.com/kata-containers/kata-containers/blob/16d6e37196cb5c18ca5e718549f7a747658105a2/src/runtime/go.mod#L118

The forked version is based on the old version of containred, that still has dependency to gogo, and has its own image API protobuf definition.
https://github.com/confidential-containers/containerd/blob/CC-main/runtime/v2/task/image.proto

**Comment by bpradipt** [comment]
cc @fidencio @bergwolf and others, is there any discussions on removal of deprecated gogo/protobuf in Kata ?

**Comment by cfircohen** [comment]
Friendly ping.  Is what Yohei proposed above a reasonable change for the Kata project?

**Comment by fidencio** [comment]
What's been proposed by @yoheiueda is reasonable, but it should first be done on `main`, and then backported to the `CCv0` branch.

**Comment by beraldoleal** [comment]
Hi all, the remaining work on "main" has been addressed in #8061. Would someone mind reviewing this PR?

CCv0 would still need to remove containerd fork, tho.

**Comment by beraldoleal** [comment]
This issue was closed, but the only item was not addressed is the CCv0 branch. We can reopen this issue or create a separated issue just for that.
